pkg/beaconfqdn/analyzer.go

```diff
@@ -0,0 +1,415 @@
+
+		if len(resList) <= 0 {
+			newFlag = true
+		}
+
+		if newFlag {
+
+			query["$push"] = bson.M{
+				"dat": bson.M{
+					"icfqdn": fqdn,
+					"icert":  1,
+					"cid":    a.chunk,
+				}}
+
+			// create selector for output
+			output.query = query
+			output.selector = src.BSONKey()
+
+		} else {
+
+			query["$set"] = bson.M{
+				"dat.$.icert": 1,
+				"dat.$.cid":   a.chunk,
+			}
+
+			// create selector for output
+			output.query = query
+			output.selector = hostSelector
+		}
+	}
+
+	return output
+}
+
+func (a *analyzer) hostBeaconQuery(score float64, src data.UniqueIP, fqdn string) updateInfo {
+	ssn := a.db.Session.Copy()
+	defer ssn.Close()
+
+	var output updateInfo
+
+	// create query
+	query := bson.M{}
+
+	// check if we need to update
+	// we do this before the other queries because otherwise if a beacon
+	// starts out with a high score which reduces over time, it will keep
+	// the incorrect high max for that specific destination.
+	var resListExactMatch []interface{}
+
+	maxBeaconMatchExactQuery := src.BSONKey()
+	maxBeaconMatchExactQuery["dat.mbfqdn"] = fqdn
+
+	_ = ssn.DB(a.db.GetSelectedDB()).C(a.conf.T.Structure.HostTable).Find(maxBeaconMatchExactQuery).All(&resListExactMatch)
+
+	// if we have exact matches, update to new score and return
+	if len(resListExactMatch) > 0 {
+		query["$set"] = bson.M{
+			"dat.$.max_beacon_fqdn_score": score,
+			"dat.$.mbfqdn":                fqdn,
+			"dat.$.cid":                   a.chunk,
+		}
+
+		// create selector for output
+		output.query = query
+
+		// using the same find query we created above will allow us to match and
+		// update the exact chunk we need to update
+		output.selector = maxBeaconMatchExactQuery
+
+		return output
+	}
+
+	// The below is only for cases where the ip is not currently listed as a max beacon
+	// for a source
+	// update max beacon score
+	newFlag := false
+	updateFlag := false
+
+	// this query will find any matching chunk that is reporting a lower
+	// max beacon score than the current one we are working with
+	maxBeaconMatchLowerQuery := src.BSONKey()
+	maxBeaconMatchLowerQuery["dat"] = bson.M{
+		"$elemMatch": bson.M{
+			"cid":                   a.chunk,
+			"max_beacon_fqdn_score": bson.M{"$lte": score},
+		},
+	}
+	// find matching lower chunks
+	var resListLower []interface{}
+
+	_ = ssn.DB(a.db.GetSelectedDB()).C(a.conf.T.Structure.HostTable).Find(maxBeaconMatchLowerQuery).All(&resListLower)
+
+	// if no matching chunks are found, we will set the new flag
+	if len(resListLower) <= 0 {
+
+		maxBeaconMatchUpperQuery := src.BSONKey()
+		maxBeaconMatchUpperQuery["dat"] = bson.M{
+			"$elemMatch": bson.M{
+				"cid":                   a.chunk,
+				"max_beacon_fqdn_score": bson.M{"$gte": score},
+			},
+		}
+
+		// find matching upper chunks
+		var resListUpper []interface{}
+		_ = ssn.DB(a.db.GetSelectedDB()).C(a.conf.T.Structure.HostTable).Find(maxBeaconMatchUpperQuery).All(&resListUpper)
```

Comment:
We don't do the logging in any of the nested inner analysis threads of the packages, just the writers, but we can plan out how and why we want to change that and update it across the board if we want.

Reply:
I'd like to create a ticket for logging more errors across the project. On the commercial side, logging the results of the mongo queries has been super helpful for tech support. 

It shouldn't be too difficult since the logger is carried along the resources struct everywhere. I think we can filter out warnings to not pop up in the console as well if that is needed in places.